spring-cloud-stream-binder-jms-activemq/src/main/java/org/springframework/cloud/stream/binder/jms/activemq/ActiveMQQueueProvisioner.java

```diff
@@ -8,34 +27,120 @@
+import org.springframework.cloud.stream.binder.jms.utils.DestinationNames;
+import org.springframework.cloud.stream.provisioning.ConsumerDestination;
+import org.springframework.cloud.stream.provisioning.ProducerDestination;
+import org.springframework.cloud.stream.provisioning.ProvisioningProvider;
 import org.springframework.jms.support.JmsUtils;
 
 /**
- * {@link QueueProvisioner} for ActiveMQ.
+ * {@link ProvisioningProvider} for ActiveMQ.
  *
  * @author José Carlos Valero
  * @since 1.1
  */
-public class ActiveMQQueueProvisioner implements QueueProvisioner{
+public class ActiveMQQueueProvisioner implements
+		ProvisioningProvider<ConsumerProperties, ProducerProperties> {
 
 	public static final String ACTIVE_MQ_DLQ = "ActiveMQ.DLQ";
+
+	private final Log logger = LogFactory.getLog(getClass());
 	private final ActiveMQConnectionFactory connectionFactory;
 
-	public ActiveMQQueueProvisioner(ActiveMQConnectionFactory connectionFactory) {
+	private final DestinationNameResolver destinationNameResolver;
+
+	public ActiveMQQueueProvisioner(ActiveMQConnectionFactory connectionFactory, DestinationNameResolver destinationNameResolver) {
 		this.connectionFactory = connectionFactory;
+		this.destinationNameResolver = destinationNameResolver;
+	}
+
+	@Override
+	public ProducerDestination provisionProducerDestination(final String name, ProducerProperties properties) {
+
+		Collection<DestinationNames> topicAndQueueNames =
+				this.destinationNameResolver.resolveTopicAndQueueNameForRequiredGroups(name, properties);
+
+		final Map<Integer, Topic> partitionTopics = new HashMap<>();
+
+		for (DestinationNames destinationNames : topicAndQueueNames) {
+			Topic topic = provisionTopic(destinationNames.getTopicName());
+			provisionConsumerGroup(destinationNames.getTopicName(),
+					destinationNames.getGroupNames());
+
+			if (destinationNames.getPartitionIndex() != null) {
+				partitionTopics.put(destinationNames.getPartitionIndex(), topic);
+			}
+			else {
+				partitionTopics.put(-1, topic);
+			}
+		}
+		return new JmsProducerDestination(partitionTopics);
 	}
 
 	@Override
-	public Destinations provisionTopicAndConsumerGroup(String topicName, String... consumerGroupName) {
+	public ConsumerDestination provisionConsumerDestination(String name, String group, ConsumerProperties properties) {
+		String groupName = this.destinationNameResolver.resolveQueueNameForInputGroup(group, properties);
+		String topicName = this.destinationNameResolver.resolveQueueNameForInputGroup(name, properties);
+
+		provisionTopic(topicName);
+		final Queue queue = provisionConsumerGroup(topicName, groupName);
+
+		//DLQ
+		Session session = null;
+		Connection connection = null;
+		try {
+			connection = connectionFactory.createConnection();
+			session = connection.createSession(true, 1);
+			session.createQueue(ACTIVE_MQ_DLQ);
+		}
+		catch (JMSException e) {
+			throw new IllegalStateException(e);
+		}
+		finally {
+			try {
+				JmsUtils.commitIfNecessary(session);
```

Comment:
This should be in its own try/catch so the closes will be reliably called; even if we get a `JMSException` on the commit.